src/lib.rs

```diff
@@ -2748,6 +2968,199 @@ mod tests {
+            Err(msgs::OVERFLOW)
+        );
+
+        // Validates that `validate_cast` satisfies its own documented safety
+        // postconditions, and also a few other properties that aren't
+        // documented but we want to guarantee anyway.
+        fn validate_behavior(
+            (layout, addr, bytes_len, cast_type): (DstLayout, usize, usize, _CastType),
+        ) {
+            if let Some((elems, split_at)) = layout._validate_cast(addr, bytes_len, cast_type) {
+                let (base_size, align, trailing_elem_size) = (
+                    layout._base_layout.size(),
+                    layout._base_layout.align(),
+                    layout._trailing_slice_elem_size,
+                );
+
+                let debug_str = format!(
+                    "layout({base_size}, {align}, {trailing_elem_size:?}).validate_cast({addr}, {bytes_len}, {cast_type:?}) => ({elems}, {split_at})",
+                );
+
+                // If this is a sized type (no trailing slice), then `elems` is
+                // meaningless, but in practice we set it to 0. Callers are not
+                // allowed to rely on this, but a lot of math is nicer if
+                // they're able to, and some callers might accidentally do that.
+                assert!(!(trailing_elem_size.is_none() && elems != 0), "{}", debug_str);
+
+                let resulting_size = base_size + (elems * trailing_elem_size.unwrap_or(0));
+                // Test that, for unsized types, `validate_cast` computed the
+                // largest possible value that fits in the given byte range.
+                assert!(
+                    trailing_elem_size
+                        .map(|elem_size| resulting_size + elem_size > bytes_len)
+                        .unwrap_or(true),
+                    "{}",
+                    debug_str
+                );
+
+                // Test safety postconditions guaranteed by `validate_cast`.
+                assert!(resulting_size <= bytes_len);
+                match cast_type {
+                    _CastType::_Prefix => {
+                        assert_eq!(addr % align, 0, "{}", debug_str);
+                        assert_eq!(resulting_size, split_at, "{}", debug_str);
+                    }
+                    _CastType::_Suffix => {
+                        assert_eq!(split_at, bytes_len - resulting_size, "{}", debug_str);
+                        assert_eq!((addr + split_at) % align, 0, "{}", debug_str);
+                    }
+                }
+            }
+        }
+
+        let layouts = itertools::iproduct!(0..8, [1, 2, 4, 8], (1..8).map(Some).chain([None]))
+            .filter(|(size, align, trailing_elem_size)| {
+                size % align == 0 && trailing_elem_size.unwrap_or(*align) % align == 0
+            })
+            .map(|(s, a, t)| layout(s, a, t));
+        itertools::iproduct!(layouts, 0..8, 0..8, [_CastType::_Prefix, _CastType::_Suffix])
+            .for_each(validate_behavior);
```

Comment:
@jswrenn Slightly refactored this after we spoke offline (including changing the name and signature of `validate_behavior`).